### PR TITLE
WIP: Implementation of two PPC plots

### PR DESCRIPTION
**Pull request by lbollar**
Referencing [#354](https://github.com/blei-lab/edward/issues/354)

Very rough WIP request. Right now, I don't have these two functions using edward methods. They are designed at the moment to use numpy arrays. I would image the workflow would look like using posterior.sample() or edward.ppc() to generate the needed data, and then passing this to the plotting functions. 

I messed with actually using edward to generate the data inside the plotting functions, but this seemed cleaner at the moment. 

[Here](https://gist.github.com/lbollar/3737d3f668882d2da050ed20017f6e6b) is a test notebook showing what these plots generate at the moment. These were meant to replicate some of the work being done in bayesplot. I just used random numbers to simulate samples in the first one as I was running a bit short on time tonight.

Thanks!

### CODE REVIEW

**Comment by dustinvtran** [comment]
Nice work. Seaborn makes sense, as well as the current structure of function arguments. I think you're right that they should work on NumPy arrays, and that the user is expected to call these after calling `ed.ppc`.  In any case, it's easy to change either way.

**Comment by dustinvtran** [comment]
Bump. How's progress on the plots? I'd be happy to get this merged in incrementally. Don't feel obligated to submit a massive slate of work at a time. For example, adding tests to these existing plots would be sufficient to merge now.

**Comment by lbollar** [comment]
Sorry Dustin, I went down the NUTS road for a bit and got busy at work, but
I will try and write tests for these and submit. See anything that needs
polishing here?

File: edward/criticisms/ppc_plots.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import seaborn as sns
```
Comment (dustinvtran) [inline edward/criticisms/ppc_plots.py]: You should add
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```
to the top of any new file.

File: edward/criticisms/ppc_plots.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def ppc_density_plot(y, y_rep):
+  """
+  Cretes 1D kernel density plot comparing data to samples from posterior.
```
Comment (dustinvtran) [inline edward/criticisms/ppc_plots.py]: typo; you could also collapse these two lines, e.g.,
```python
"""Creates 1D ...
```
Same goes for docstring below.

File: edward/criticisms/ppc_plots.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  Parameters
+  ----------
+  y : numpy array
```
Comment (dustinvtran) [inline edward/criticisms/ppc_plots.py]: `numpy array` -> `np.ndarray`

File: edward/criticisms/ppc_plots.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  Parameters
+  ----------
+  y_stats : numpy array
+    A 1D numpy array
+  yrep_stats : float
+    Float representing statistic value of observed data
+  stat_name : string
+    Optional string value for including statistic name in legend
+  Keyword Arguments
```
Comment (dustinvtran) [inline edward/criticisms/ppc_plots.py]: `Keyword Arguments` -> `**kwargs`, to stay consistent with other Edward docstrings

File: edward/criticisms/ppc_plots.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  """
+  Cretes histogram plot comparing data to samples from posterior.
+
+  Parameters
+  ----------
+  y_stats : numpy array
+    A 1D numpy array
+  yrep_stats : float
+    Float representing statistic value of observed data
+  stat_name : string
```
Comment (dustinvtran) [inline edward/criticisms/ppc_plots.py]: `string, optional`

File: edward/criticisms/ppc_plots.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import seaborn as sns
```
Comment (dustinvtran) [inline edward/criticisms/ppc_plots.py]: This adds an install requirement to Edward. I prefer minimal requirements, so can you wrap this in a try/except as follows?
```python
try:
  import seaborn as sns
except ImportError:
  pass
```

**Comment by lbollar** [comment]
Looking at Travis, do i need to do the same thing for matplotlib in tests?

File: tests/test-criticisms/test_ppc_plots.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import matplotlib
+matplotlib.use('Agg', warn=False)
```
Comment (lbollar) [inline tests/test-criticisms/test_ppc_plots.py]: I wasn't sure how necessary this line is, but I didn't know much about testing plotting functions, and so I took a look at what PyMC3 had done and took this from there.

Comment (dustinvtran) [inline tests/test-criticisms/test_ppc_plots.py]: Got it. I think it's not needed. The aim of the test is just to show that the function successfully terminates.

**Comment by dustinvtran** [comment]
Good catch; following above comment, no, but you should definitely add seaborn to `.travis.yml`. For example, after line 32,
```python
  - pip install seaborn
```

**Comment by lbollar** [comment]
Looks like build failed for 2.7, and it looks display related maybe. Think that could possibly be related to what pycharm had in their test code relating to choosing a dislay.

**Comment by dustinvtran** [comment]
I was able to solve this problem. Not sure how to add commits to your PR; can you add a commit making these exact changes? https://github.com/blei-lab/edward/commit/8983355215b57e248c03afaa71b907a3b7ae57fc

**Comment by lbollar** [comment]
Sorry I haven't been able to figure this out. Looks like the same error on 2.7 still. Did I miss something?

**Comment by lbollar** [comment]
I saw the addition of a sleep command [here](http://stackoverflow.com/questions/35403127/testing-matplotlib-based-plots-in-travis-ci).

```yml
before_script: # configure a headless display to test plot generation
- "export DISPLAY=:99.0"
- "sh -e /etc/init.d/xvfb start"
- sleep 3 # give xvfb some time to start
```
Not sure if that would do anything to fix this or not...

**Comment by dustinvtran** [comment]
It seems to work on my local branch? I'll merge anyways. It might be a permission problem; if not, I'll look into it again.

**Pull request merged**